components/script/dom/gpubuffer.rs

```diff
@@ -313,24 +314,21 @@ impl GPUBufferMethods for GPUBuffer {
             return Err(Error::Operation);
         }
 
-        unsafe extern "C" fn free_func(_contents: *mut c_void, free_user_data: *mut c_void) {
-            let _ = Rc::from_raw(free_user_data as _);
-        }
+        let vec = m_info.mapping.borrow().to_vec();
```

Comment:
Here the data is copied, but we need to create an array buffer using a mutable pointer to the data. 

We should add a new util in `typedarrays.rs` that wraps the call to `NewExternalArrayBuffer`, and which returns an array buffer not initialized with a copy of the data, but rather one that will have its internal object set to that returned by the call to `NewExternalArrayBuffer`(a bit like you are doing below with `new`). 

Regarding `free_func` and `free_user_data`, their current use appears useless to me: `free_func` seems to free a raw pointer to the clone of the mapping, passed as `free_user_data` to `NewExternalArrayBuffer`.

Looking at how Gecko [uses it](https://searchfox.org/mozilla-central/rev/c88ac92342d525fd69dbd4c29cd22984eaf9de78/dom/webgpu/Buffer.cpp#255) I would suggest the following workflow:

1. In `Unmap`, detach the array buffer. 
2. Wait for `free_func` to be called, and only then send `WebGPURequest::UnmapBuffer` to the backend and drop the locally held mapping. This requires passing `GPUBufferMapInfo` as the `free_user_data` raw pointer. 

By the way, I am wondering if we are really sharing memory over IPC, or rather just cloning data around. Does the current implementation work @sagudev ?

Perhaps the freeing of resources should be further researched and left as a follow-up. 

@jdm Does this make sense?